packages/timeline-state-resolver/src/integrations/vmix/vMixResponseStreamReader.ts

```diff
@@ -0,0 +1,104 @@
+import { EventEmitter } from 'eventemitter3'
+
+type ResponseStreamReaderEvents = {
+	response: [response: Response]
+	error: [error: Error]
+}
+
+export interface Response {
+	command: string
+	response: 'OK' | 'ER'
+	message: string
+	body?: string
+}
+
+const RESPONSE_REGEX = /^(?<command>\w+)\s+(?<response>OK|ER|\d+)(\s+(?<responseMsg>.*))?/i
+
+/**
+ * A receiver for vMix responses
+ */
+export class VMixResponseStreamReader extends EventEmitter<ResponseStreamReaderEvents> {
+	private _unprocessedLines: string[] = []
+	private _lineRemainder = ''
+
+	reset() {
+		this._unprocessedLines = []
+		this._lineRemainder = ''
+	}
+
+	processIncomingData(data: Buffer) {
+		const string = this._lineRemainder + data.toString('utf-8')
+		this._lineRemainder = ''
+		const lines = string.split('\r\n')
+
+		if (lines[lines.length - 1] === '') {
+			// the data ends with a newline
+			this._unprocessedLines.push(...lines)
+		} else {
+			const incompleteLine = lines.pop()
+			this._unprocessedLines.push(...lines)
+			// we need to keep the remaining incomplete line
+			if (incompleteLine != null) {
+				this._lineRemainder = incompleteLine
+			}
+		}
+
+		let lineToProcess: string | undefined
+
+		while ((lineToProcess = this._unprocessedLines.shift()) != null) {
+			const result = RESPONSE_REGEX.exec(lineToProcess)
+
+			if (result && result.groups?.['response']) {
+				const responseLen = parseInt(result?.groups?.['response'])
+
+				// create a response object
+				const response: Response = {
+					command: result.groups?.['command'],
+					response: (Number.isNaN(responseLen) ? result.groups?.['response'] : 'OK') as Response['response'],
+					message: result.groups?.['responseMsg'],
+					body: undefined as undefined | string,
+				}
+
+				// parse payload data if there is any
+				if (!Number.isNaN(responseLen)) {
+					const payloadData = this.processPayloadData(responseLen)
+					if (payloadData == null) {
+						this._unprocessedLines.unshift(lineToProcess)
+						break
+					} else {
+						response.body = payloadData
+					}
+				}
+
+				// now do something with response
+				try {
```

Comment:
I feel like it would be better to put this `try` above the `const response: Response` (or perhaps just inside of the `while`?) so that if there are other bugs here it wont break the loop, and will end up ignoring a message instead of getting stuck again

Reply:
I think that would be going too far in not trusting our own code.

Reply:
It is a bit, but I still like doing it. I don't think it makes the code any less readable and helps reinforce the approach of 'if something errors ignore it and continue'.
Im not opposed to leaving it as is though as the real world benefit is likely non existent

Reply:
Updated as requested